vignettes/decorate-module-output.Rmd

```diff
@@ -0,0 +1,482 @@
+`server` components, where they will be used by `ui/srv_teal_transform_module`. 
+
+Please find an example module for the sake of this article:
+
+
+```{r}
+tm_decorated_plot <- function(label = "module", transformators = list(), decorators = teal_transform_module()) {
+  module(
+    label = label,
+    ui = function(id, decorators) {
+      ns <- NS(id)
+      div(
+        selectInput(ns("dataname"), label = "select dataname", choices = NULL),
+        selectInput(ns("x"), label = "select x", choices = NULL),
+        selectInput(ns("y"), label = "select y", choices = NULL),
+        ui_teal_transform_data(ns("decorate"), transformators = decorators),
+        plotOutput(ns("plot")),
+        verbatimTextOutput(ns("text"))
+      )
+    },
+    server = function(id, data, decorators) {
+      moduleServer(id, function(input, output, session) {
+        observeEvent(data(), {
+          updateSelectInput(inputId = "dataname", choices = names(data()))
+        })
+
+        observeEvent(input$dataname, {
+          req(input$dataname)
+          updateSelectInput(inputId = "x", choices = colnames(data()[[input$dataname]]))
+          updateSelectInput(inputId = "y", choices = colnames(data()[[input$dataname]]))
+        })
+
+        plot_data <- debounce(
+          reactive({
+            req(input$dataname, input$x, input$y)
+            within(data(),
+              {
+                plot <- ggplot2::ggplot(dataname, ggplot2::aes(x = x, y = y)) +
+                  ggplot2::geom_point()
+              },
+              dataname = as.name(input$dataname),
+              x = as.name(input$x),
+              y = as.name(input$y)
+            )
+          }), 200
+        )
+
+        plot_data_decorated <- srv_teal_transform_data("decorate", data = plot_data, transformators = decorators)
+
+        plot_r <- reactive({
+          req(plot_data_decorated())
+          plot_data_decorated()[["plot"]]
+        })
+
+        output$plot <- renderPlot(plot_r())
+        output$text <- renderText({
+          teal.code::get_code(req(plot_data_decorated()))
+        })
+      })
+    },
+    ui_args = list(decorators = decorators),
+    server_args = list(decorators = decorators)
+  )
+}
+```
+
+### Application
+
+```{r}
+library(ggplot2)
+app <- init(
+  data = teal_data(iris = iris, mtcars = mtcars),
+  modules = modules(
+    tm_decorated_plot("identity"),
+    tm_decorated_plot("no-ui", decorators = static_decorator),
+    tm_decorated_plot("lang", decorators = static_decorator_lang),
+    tm_decorated_plot("interactive", decorators = interactive_decorator),
+    tm_decorated_plot("interactive-from lang", decorators = interactive_decorator_lang),
+    tm_decorated_plot("from-fun", decorators = gg_xlab_decorator("plot")),
+    tm_decorated_plot("failing", decorators = failing_decorator)
+  )
+)
+
+if (interactive()) {
+  shinyApp(app$ui, app$server)
+}
+```
+
+## Multiple Decorators
```

Comment:
can we have more general example of multiple (n) decorators?
the current one is for n=2 decorators - not 1, not 0, not 3, etc. - exactly 2

Reply:
I tried to rewrite the code so it works for n decorators. here is the PR, however it is not done yet as this does not produce the final plot https://github.com/insightsengineering/teal/pull/1418

Reply:
Ok, managed to fix the reactivity and merged the vignette to this PR in here https://github.com/insightsengineering/teal/pull/1357/commits/42035b7877c2da21bd2c616ed8d7f7c3dc931482